File: stella_invicta_extreme_edition/lib/database/systems/date.ex
defmodule StellaInvicta.System.Date do
  @behaviour StellaInvicta.System

  alias StellaInvicta.MessageQueue

  @impl true
  def run(world) do
    # Get current date or default
    current_date = Map.get(world, :date, %{hour: 0, day: 1, month: 1, year: 1})

    # Calculate new date using the helper function
    {new_date, events} = advance_time(current_date)

    # Publish any date events
    world =
      Enum.reduce(events, world, fn event, acc ->
        MessageQueue.publish(acc, :date_events, event)
      end)

    # Update the world
    Map.put(world, :date, new_date)
  end

  @impl true
  def subscriptions, do: []

  @impl true
  def handle_message(world, _topic, _message), do: world

  # 1. End of Year: Month >= 12, Day >= 30, Hour >= 23
  defp advance_time(%{month: month, day: day, hour: hour} = date)
       when month >= 12 and day >= 30 and hour >= 23 do
    new_date = %{date | hour: 0, day: 1, month: 1, year: date.year + 1}
    events = [:new_hour, :new_day, :new_month, {:new_year, new_date.year}]
    {new_date, events}
  end

  # 2. End of Month: Any Month, Day >= 30, Hour >= 23
  defp advance_time(%{day: day, hour: hour} = date)
       when day >= 30 and hour >= 23 do
    new_date = %{date | hour: 0, day: 1, month: date.month + 1}
    events = [:new_hour, :new_day, {:new_month, new_date.month}]
    {new_date, events}
  end

  # 3. End of Day: Any Day, Hour >= 23
  defp advance_time(%{hour: hour} = date)
       when hour >= 23 do
    new_date = %{date | hour: 0, day: date.day + 1}
    events = [:new_hour, {:new_day, new_date.day}]
    {new_date, events}
  end

  # 4. Normal Hour: Any other time
  # Result: Just Hour + 1
  defp advance_time(date) do
    new_date = %{date | hour: date.hour + 1}
    {new_date, [:new_hour]}
  end
end

File: InvictaDB/StellaInvicta/System/Date.cs
using InvictaDB;
using InvictaDB.Messaging;
using static InvictaDB.Messaging.GameMessages;

namespace StellaInvicta.System.Date;

/// <summary>
/// Manages in-game date and time.
/// </summary>
public class DateSystem : ISystem
{
    /// <inheritdoc/>
    public string Name => DateSystemSender;

    /// <inheritdoc/>
    public string Description => "Manages in-game date and time.";
    /// <inheritdoc/>
    public string Author => "InvictaDB Team";
    /// <inheritdoc/>
    public bool Enabled { get; set; } = true;
    /// <summary>
    /// Indicates whether the system has been initialized.
    /// </summary>
    public bool IsInitialized { get; set; }

    /// <inheritdoc/>
    public InvictaDatabase Initialize(InvictaDatabase db)
    {
        IsInitialized = true;
        return db.InsertSingleton(new DateTime(1, 1, 1, 0, 0, 0));
    }
    /// <inheritdoc/>
    public InvictaDatabase Shutdown(InvictaDatabase db)
    {
        // No specific shutdown logic needed for the date system.
        return db;
    }
    /// <summary>
    /// Advances the in-game date by one hour.
    /// </summary>
    /// <param name="db"></param>
    /// <returns></returns>
    public InvictaDatabase Run(InvictaDatabase db)
    {
        // Get the current game date
        var currentDate = db.GetSingleton<DateTime>();

        // Advance the date by one hour
        var newDate = currentDate.AddHours(1);

        // Update the database with the new date
        db = db.InsertSingleton(newDate);

        // Send time-related messages
        db = SendTimeMessages(db, currentDate, newDate);

        return db;
    }

    /// <summary>
    /// Sends messages for any time boundaries that were crossed.
    /// </summary>
    private InvictaDatabase SendTimeMessages(InvictaDatabase db, DateTime oldDate, DateTime newDate)
    {
        // Always send NewHour message
        db = db.SendMessage(Name, new NewHour(newDate, oldDate));

        // Check if we crossed into a new day
        if (newDate.Date != oldDate.Date)
        {
            db = db.SendMessage(Name, new NewDay(newDate, oldDate));

            // Check for new week (Monday)
            if (newDate.DayOfWeek == DayOfWeek.Monday)
            {
                var weekNumber = GetWeekOfYear(newDate);
                db = db.SendMessage(Name, new NewWeek(newDate, weekNumber));
            }
        }

        // Check if we crossed into a new month
        if (newDate.Month != oldDate.Month)
        {
            db = db.SendMessage(Name, new NewMonth(newDate, oldDate.Month));

            // Check for season change (months 3, 6, 9, 12)
            var oldSeason = GetSeason(oldDate);
            var newSeason = GetSeason(newDate);
            if (newSeason != oldSeason)
            {
                db = db.SendMessage(Name, new NewSeason(newDate, newSeason));
            }
        }

        // Check if we crossed into a new year
        if (newDate.Year != oldDate.Year)
        {
            db = db.SendMessage(Name, new NewYear(newDate, oldDate.Year));
        }

        return db;
    }

    /// <summary>
    /// Gets the week number of the year for a given date.
    /// </summary>
    private static int GetWeekOfYear(DateTime date)
    {
        var cal = global::System.Globalization.CultureInfo.InvariantCulture.Calendar;
        return cal.GetWeekOfYear(date, global::System.Globalization.CalendarWeekRule.FirstFourDayWeek, DayOfWeek.Monday);
    }
}